README.md

```diff
@@ -1 +1,471 @@
+parsing just to work with binary data.
+
+### GET ping
+
+There is _one_ GET endpoint, used for debugging, testing and health checks.
+
+All GET requests to the bootstrap service will receive the response string `OK`
+encoded as UTF-8 text data and the `200` status code.
+
+The GET endpoint behaves differently to all other methods in that it is not
+serialized, is not binary data, has no op headers and cannot interact with the
+kv service at all.
+
+### MessagePack serialization
+
+All requests and responses are serialized as binary MessagePack data in the
+body of the request/response.
+
+The op header (see below) defines which operation the messagepack payload in a
+request is dispatched to and what kind of response will be returned.
+
+This is the same serialization format that holochain itself uses for wire
+messages on the network and for compatibility with wasm for validation logic.
+
+### Ed25519/NaCl crytography
+
+All cryptographic logic is handled as per libsodium using the Ed25519 curve.
+
+This is the same cryptography as holochain itself which means the signatures and
+validation used by the bootstrap service are the same as those used by the agent
+DHT by conductors.
+
+This implementation uses [tweetnacl](https://www.npmjs.com/package/tweetnacl).
+
+### Headers
+
+The `Content-Type` header for all POST requests must be `application/octet` to
+signify to the server that the body of the request is a binary payload.
+
+The action to be performed is set by the `X-Op` header in the POST request.
+
+The possible values are:
+
+- `put`: store signed agent info
+- `list`: list all stored agent info
+- `get`: retrive a single agent info
+- `random`: retrieve up to N random agents (hybrid of list and get)
+
+## Data structures
+
+### Agent info
+
+`AgentInfoSigned` is the main data structure.
+
+It is the same structure defined by `kitsune_p2p` in the Rust codebase for the
+conductor, but ported to typescript for validation here.
+
+It looks like this on the wire:
+
+```typescript
+{
+ signature: Uint8Array,
+ agent: Uint8Array,
+ agent_info: Uint8Array,
+}
+```
+
+Where the `agent_info` is messagepack serialized binary data that MUST be valid
+for the `agent` public key and `signature` bytes according to libsodium.
+
+If the `agent_info` is not valid then it MUST be discarded and any further logic
+abandoned because this is ALWAYS malicious or corrupt data. The inner
+`agent_info` MUST NOT be deserialized if it is invalid.
+
+When the `agent_info` is validated and unpacked it looks like this:
+
+```typescript
+agent_info: {
+ space: Uint8Array,
+ agent: Uint8Array,
+ urls: Array<string>,
+ signed_at_ms: number,
+}
+```
+
+- `space` is the bytes of the hash of the DNA used to connect to the DHT
+- `agent` is the same as the signing key above and MUST match it
+- `urls` is an array of strings that are the URLs the agent can be found at
+- `signed_at_ms` is the unix millisecond timestamp of the signing
```

Comment:
similar to above, do we want to give details about the structure of space / agent?